### DISCUSSION TITLE
Infer return type for non-generic async method builder from SetResult

### DISCUSSION THREAD

**Discussion by sharwell** [dotnet/csharplang#715] [General]
Currently a task-like type with a non-generic `AsyncMethodBuilder` requires the `SetResult` method to not take an argument (the asynchronous equivalent of a `void` method). It would be helpful if the effective return type for such a case was inferred from the argument, if any, accepted by the `SetResult` method. For example, suppose you had the following:

```csharp
[AsyncMethodBuilder(typeof(IntTaskMethodBuilder))]
public struct IntTask {
  // ...
}

public struct IntTaskMethodBuilder {
  // ...

  public void SetResult(int result) // <-- if this was allowed
  {
  }
}

// Then we could do this:
public async IntTask MethodAsync()
{
  return 1;
}
```

**Comment by paulomorgado** [comment]
@sharwell, I understand what you're asking for and it can be done. But, what's your motivation for this?

Would multiple overloads of `SetResult` allow the task method builder to be reusable across multiple task-like types?

**Comment by sharwell** [comment]
@paulomorgado it came up during an experiment I'm working on. Right now it's so early that I can't say for sure whether or not I would use this. I have a working prototype, but haven't put it into practice enough to see how it holds up.

> Would multiple overloads of `SetResult` allow the task method builder to be reusable across multiple task-like types?

This seems lower priority but reasonable.
